projects/07/VMTranslator.py

```diff
@@ -0,0 +1,346 @@
+        self.write_line('D=M')
+        self.write_decrement_sp()
+        self.write_line('D=M-D')
+        self.write_line('@{label}'.format(label=CASE_TRUE))
+
+    def write_push_true_false(self, CASE_DONE, CASE_TRUE):
+        self.write_line('@0')
+        self.write_line('D=A')
+        self.write_line('@{label}'.format(label=CASE_DONE))
+        self.write_line('0;JMP')
+        self.write_line('({label})'.format(label=CASE_TRUE))
+        self.write_line('@0')
+        self.write_line('D=A-1')
+        self.write_line('({label})'.format(label=CASE_DONE))
+        self.write_line('@SP')
+        self.write_line('A=M')
+        self.write_line('M=D')
+
+    def write_increment_sp(self):
+        '''Update the stack pointer assuming a push occurred'''
+        self.write_line('@SP')
+        self.write_line('M=M+1')
+
+    def write_decrement_sp(self):
+        '''Update the stack pointer assuming a pop occurred. M contains the popped value'''
+        self.write_line('@SP')
+        self.write_line('AM=M-1')
+
+    def write_push_pop(self, cmd_type, segment, index):
+        '''
+        Writes to the output file the assembly code that implements the given push/pop command
+        '''
+        if cmd_type == CmdType.C_PUSH:
+            self.write_push(segment, index)
+        if cmd_type == CmdType.C_POP:
+            self.write_pop(segment, index)
+
+    def write_push(self, segment, index):
+        if segment == CONSTANT:
+            self.write_line('@{index}'.format(index=index))
+            self.write_line('D=A')
+        elif segment in [LOCAL, ARGUMENT, THIS, THAT]:
+            self.offset_segment(segment, index)
+            self.write_line('A=D')
+            self.write_line('D=M')
+        elif segment in [POINTER, TEMP]:
+            self.offset_literal(segment, index)
+            self.write_line('A=D')
+            self.write_line('D=M')
+        else:
+            #static
+            static_label=self.generate_label(index)
+            self.write_line('@{label}'.format(label=static_label))
+            self.write_line('D=M')
+        self.write_line('@SP')
+        self.write_line('A=M')
+        self.write_line('M=D')
+        self.write_increment_sp()
+
+    def write_pop(self, segment, index):
+        if segment in [LOCAL, ARGUMENT, THIS, THAT]:
+            self.offset_segment(segment, index)
+        elif segment in [POINTER, TEMP]:
+            self.offset_literal(segment, index)
+        else:
+            #static
+            static_label = self.generate_label(index)
+            self.write_line('@{label}'.format(label=static_label))
+            self.write_line('D=A')
+        self.write_line('@SP')
+        self.write_line('A=M')
+        self.write_line('M=D')
+        self.write_decrement_sp()
+        self.write_line('D=M')
+        self.write_line('@SP')
+        self.write_line('A=M+1')
+        self.write_line('A=M')
+        self.write_line('M=D')
+
+    def offset_literal(self, segment, index):
```

Comment:
offset_literal and offset_segment do very nearly the same thing. I might want to combine the common elements and just have setting the D register in these methods

Reply:
#6 tracks this